File: packages/react-native-client/android/src/main/java/io/fishjam/reactnative/VideoRendererViewModule.kt
package io.fishjam.reactnative

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class VideoRendererViewModule : Module() {
  override fun definition() =
    ModuleDefinition {
      Name("VideoRendererViewModule")

      Events("TrackAspectRatioUpdated")

      View(VideoRendererView::class) {
        OnViewDestroys { view: VideoRendererView ->
          view.dispose()
          RNFishjamClient.trackUpdateListenersManager.remove(view)
          RNFishjamClient.localTracksSwitchListenerManager.remove(view)
        }

        Prop("trackId") { view, trackId: String ->
          view.init(trackId)
        }

        Prop("videoLayout") { view, videoLayout: String ->
          view.setVideoLayout(videoLayout)
        }

        Prop("skipRenderOutsideVisibleArea") { view, skipRenderOutsideVisibleArea: Boolean ->
          view.checkVisibilityDelayMillis = if (skipRenderOutsideVisibleArea) 1000 else null
        }
      }
    }
}

File: packages/react-native-client/ios/VideoRendererViewModule.swift
import ExpoModulesCore

public class VideoRendererViewModule: Module {
    public func definition() -> ModuleDefinition {
        Name("VideoRendererViewModule")

        View(VideoRendererView.self) {
            Events("onDimensionsChanged")
            
            Prop("trackId") { (view, trackId) in
                view.trackId = trackId
            }

            Prop("videoLayout") { (view, videoLayout) in
                view.videoLayout = videoLayout
            }

            Prop("skipRenderOutsideVisibleArea") { (view, skipRenderOutsideVisibleArea) in
                view.checkVisibilityTimeInterval = skipRenderOutsideVisibleArea ? 1 : nil
            }
            
            AsyncFunction("getCurrentDimensions") { (renderView: VideoRendererView) -> [String: Int32] in
                [
                    "width": renderView.videoView.dimensions?.width ?? 0,
                    "height": renderView.videoView.dimensions?.height ?? 0
                ]
            }
        }
    }
}
